File: src/shiny_hunter/gbfont.py
"""Pokemon Red/Blue bitmap font renderer.

Loads the 8x8 1bpp font tiles from pokered's font.png and renders text
with the same look as in-game dialog boxes.

font.png covers tile indices $80-$FF (A-Z, a-z, digits, punctuation).
font_extra.png covers tile indices $60-$7F (border tiles, space, bold chars).
"""
from __future__ import annotations

from pathlib import Path

from PIL import Image

TILE_SIZE = 8
FONT_COLS = 16
GB_SCREEN_TILES_W = 20  # 160px / 8px per tile

_DATA_DIR = Path(__file__).parent / "data"
_FONT_PNG = _DATA_DIR / "pokered_font.png"
_FONT_EXTRA_PNG = _DATA_DIR / "pokered_font_extra.png"
_SHINY_ICON_PNG = _DATA_DIR / "shiny_icon.png"

SHINY_CHAR = "⁂"  # ⁂ — same placeholder as pokecrystal

_CHARMAP: dict[str, int] = {
    "A": 0x80, "B": 0x81, "C": 0x82, "D": 0x83, "E": 0x84,
    "F": 0x85, "G": 0x86, "H": 0x87, "I": 0x88, "J": 0x89,
    "K": 0x8A, "L": 0x8B, "M": 0x8C, "N": 0x8D, "O": 0x8E,
    "P": 0x8F, "Q": 0x90, "R": 0x91, "S": 0x92, "T": 0x93,
    "U": 0x94, "V": 0x95, "W": 0x96, "X": 0x97, "Y": 0x98,
    "Z": 0x99,
    "(": 0x9A, ")": 0x9B, ":": 0x9C, ";": 0x9D, "[": 0x9E, "]": 0x9F,
    "a": 0xA0, "b": 0xA1, "c": 0xA2, "d": 0xA3, "e": 0xA4,
    "f": 0xA5, "g": 0xA6, "h": 0xA7, "i": 0xA8, "j": 0xA9,
    "k": 0xAA, "l": 0xAB, "m": 0xAC, "n": 0xAD, "o": 0xAE,
    "p": 0xAF, "q": 0xB0, "r": 0xB1, "s": 0xB2, "t": 0xB3,
    "u": 0xB4, "v": 0xB5, "w": 0xB6, "x": 0xB7, "y": 0xB8,
    "z": 0xB9,
    "'": 0xE0,
    "-": 0xE3,
    "?": 0xE6, "!": 0xE7, ".": 0xE8,
    "/": 0xF3, ",": 0xF4,
    "0": 0xF6, "1": 0xF7, "2": 0xF8, "3": 0xF9, "4": 0xFA,
    "5": 0xFB, "6": 0xFC, "7": 0xFD, "8": 0xFE, "9": 0xFF,
    " ": 0x7F,
    "=": 0xE3,  # reuse hyphen tile as "=" stand-in
    SHINY_CHAR: 0x3F,  # Crystal's shiny sparkle icon
}

_BORDER_TL = 0x79  # ┌
_BORDER_H = 0x7A   # ─
_BORDER_TR = 0x7B  # ┐
_BORDER_V = 0x7C   # │
_BORDER_BL = 0x7D  # └
_BORDER_BR = 0x7E  # ┘

_tiles: dict[int, Image.Image] | None = None


def _load_tiles() -> dict[int, Image.Image]:
    global _tiles
    if _tiles is not None:
        return _tiles
    _tiles = {}

    # font.png: tiles $80-$FF
    sheet = Image.open(_FONT_PNG).convert("L")
    n_tiles = (sheet.width // TILE_SIZE) * (sheet.height // TILE_SIZE)
    for i in range(n_tiles):
        col = i % FONT_COLS
        row = i // FONT_COLS
        x, y = col * TILE_SIZE, row * TILE_SIZE
        _tiles[0x80 + i] = sheet.crop((x, y, x + TILE_SIZE, y + TILE_SIZE))

    # font_extra.png: tiles $60-$7F
    extra = Image.open(_FONT_EXTRA_PNG).convert("L")
    n_extra = (extra.width // TILE_SIZE) * (extra.height // TILE_SIZE)
    for i in range(n_extra):
        col = i % FONT_COLS
        row = i // FONT_COLS
        x, y = col * TILE_SIZE, row * TILE_SIZE
        _tiles[0x60 + i] = extra.crop((x, y, x + TILE_SIZE, y + TILE_SIZE))

    # shiny_icon.png: Crystal's shiny sparkle at tile $3F
    shiny = Image.open(_SHINY_ICON_PNG).convert("L")
    _tiles[0x3F] = shiny.crop((0, 0, TILE_SIZE, TILE_SIZE))

    return _tiles


def _get_tile(idx: int) -> Image.Image:
    tiles = _load_tiles()
    return tiles.get(idx, tiles.get(0x7F, Image.new("L", (TILE_SIZE, TILE_SIZE), 255)))


def render_textbox(lines: list[str], width_tiles: int | None = None) -> Image.Image:
    """Render lines of text inside a Pokemon-style bordered text box.

    Returns a grayscale image (white background, dark text/border).
    """
    max_len = max((len(line) for line in lines), default=0)
    if width_tiles is None:
        width_tiles = max_len + 4  # content + 1 padding each side + 2 borders
    total_w = width_tiles
    inner_w = total_w - 2
    total_h = len(lines) + 2

    img = Image.new("L", (total_w * TILE_SIZE, total_h * TILE_SIZE), 255)

    def paste(tile_idx: int, tx: int, ty: int) -> None:
        img.paste(_get_tile(tile_idx), (tx * TILE_SIZE, ty * TILE_SIZE))

    paste(_BORDER_TL, 0, 0)
    for x in range(1, total_w - 1):
        paste(_BORDER_H, x, 0)
    paste(_BORDER_TR, total_w - 1, 0)

    for row_i, line in enumerate(lines):
        y = row_i + 1
        paste(_BORDER_V, 0, y)
        for x in range(1, total_w - 1):
            paste(0x7F, x, y)
        for ci, ch in enumerate(line):
            idx = _CHARMAP.get(ch, 0x7F)
            paste(idx, ci + 1, y)
        paste(_BORDER_V, total_w - 1, y)

    paste(_BORDER_BL, 0, total_h - 1)
    for x in range(1, total_w - 1):
        paste(_BORDER_H, x, total_h - 1)
    paste(_BORDER_BR, total_w - 1, total_h - 1)

    return img

File: web/src/lib/gbfont.ts
/**
 * Pokemon Red/Blue bitmap font renderer for the browser.
 *
 * Loads three 1bpp tile sheets shipped under /public/gbfont/:
 *   - pokered_font.png       — tiles 0x80-0xFF (A-Z, a-z, digits, punctuation)
 *   - pokered_font_extra.png — tiles 0x60-0x7F (border tiles, space, bold)
 *   - shiny_icon.png         — Crystal's shiny sparkle, mapped to tile 0x3F
 *
 * Mirrors `src/shiny_hunter/gbfont.py` so the web monitor textbox looks
 * identical to the Python one.
 */

const TILE_SIZE = 8;
const FONT_COLS = 16;
export const GB_SCREEN_TILES_W = 20;

export const SHINY_CHAR = '⁂'; // ⁂

const CHARMAP: Record<string, number> = {
  A: 0x80, B: 0x81, C: 0x82, D: 0x83, E: 0x84,
  F: 0x85, G: 0x86, H: 0x87, I: 0x88, J: 0x89,
  K: 0x8a, L: 0x8b, M: 0x8c, N: 0x8d, O: 0x8e,
  P: 0x8f, Q: 0x90, R: 0x91, S: 0x92, T: 0x93,
  U: 0x94, V: 0x95, W: 0x96, X: 0x97, Y: 0x98,
  Z: 0x99,
  '(': 0x9a, ')': 0x9b, ':': 0x9c, ';': 0x9d, '[': 0x9e, ']': 0x9f,
  a: 0xa0, b: 0xa1, c: 0xa2, d: 0xa3, e: 0xa4,
  f: 0xa5, g: 0xa6, h: 0xa7, i: 0xa8, j: 0xa9,
  k: 0xaa, l: 0xab, m: 0xac, n: 0xad, o: 0xae,
  p: 0xaf, q: 0xb0, r: 0xb1, s: 0xb2, t: 0xb3,
  u: 0xb4, v: 0xb5, w: 0xb6, x: 0xb7, y: 0xb8,
  z: 0xb9,
  "'": 0xe0,
  '-': 0xe3,
  '?': 0xe6, '!': 0xe7, '.': 0xe8,
  '/': 0xf3, ',': 0xf4,
  '0': 0xf6, '1': 0xf7, '2': 0xf8, '3': 0xf9, '4': 0xfa,
  '5': 0xfb, '6': 0xfc, '7': 0xfd, '8': 0xfe, '9': 0xff,
  ' ': 0x7f,
  '=': 0xe3,
  [SHINY_CHAR]: 0x3f,
};

const BORDER_TL = 0x79;
const BORDER_H = 0x7a;
const BORDER_TR = 0x7b;
const BORDER_V = 0x7c;
const BORDER_BL = 0x7d;
const BORDER_BR = 0x7e;

interface Tile {
  /** Length 64; each entry is the grayscale byte (0-255). */
  pixels: Uint8ClampedArray;
}

let tilesPromise: Promise<Map<number, Tile>> | null = null;

async function loadImage(src: string): Promise<HTMLImageElement> {
  return new Promise((resolve, reject) => {
    const img = new Image();
    img.onload = () => resolve(img);
    img.onerror = (e) => reject(e);
    img.src = src;
  });
}

function imageToTiles(
  img: HTMLImageElement,
  startTileIdx: number,
  out: Map<number, Tile>,
): void {
  const cols = Math.floor(img.width / TILE_SIZE);
  const rows = Math.floor(img.height / TILE_SIZE);
  const canvas = document.createElement('canvas');
  canvas.width = img.width;
  canvas.height = img.height;
  const ctx = canvas.getContext('2d');
  if (!ctx) throw new Error('2d context unavailable');
  ctx.drawImage(img, 0, 0);
  const data = ctx.getImageData(0, 0, img.width, img.height).data;

  for (let i = 0; i < cols * rows; i++) {
    const tx = (i % FONT_COLS) * TILE_SIZE;
    const ty = Math.floor(i / FONT_COLS) * TILE_SIZE;
    const pixels = new Uint8ClampedArray(TILE_SIZE * TILE_SIZE);
    for (let y = 0; y < TILE_SIZE; y++) {
      for (let x = 0; x < TILE_SIZE; x++) {
        const srcIdx = ((ty + y) * img.width + (tx + x)) * 4;
        // Read luminance from the red channel (image is grayscale).
        pixels[y * TILE_SIZE + x] = data[srcIdx];
      }
    }
    out.set(startTileIdx + i, { pixels });
  }
}

async function loadTiles(): Promise<Map<number, Tile>> {
  if (tilesPromise) return tilesPromise;
  tilesPromise = (async () => {
    const tiles = new Map<number, Tile>();
    const [main, extra, shiny] = await Promise.all([
      loadImage('/gbfont/pokered_font.png'),
      loadImage('/gbfont/pokered_font_extra.png'),
      loadImage('/gbfont/shiny_icon.png'),
    ]);
    imageToTiles(main, 0x80, tiles);
    imageToTiles(extra, 0x60, tiles);
    imageToTiles(shiny, 0x3f, tiles);
    return tiles;
  })();
  return tilesPromise;
}

/** Pre-load the font assets. Resolves once tiles are decoded and cached. */
export function preloadFont(): Promise<void> {
  return loadTiles().then(() => undefined);
}

function getTile(tiles: Map<number, Tile>, idx: number): Tile {
  return tiles.get(idx) ?? tiles.get(0x7f)!;
}

/**
 * Render lines of text inside a Pokémon-style bordered text box.
 *
 * Returns an offscreen canvas at unscaled tile resolution: width is
 * `widthTiles * 8` pixels, height is `(lines.length + 2) * 8` pixels.
 * Pixels are RGB grayscale (0 = black, 255 = white). The caller can scale
 * with NEAREST sampling and composite over a screen.
 */
export async function renderTextbox(
  lines: string[],
  widthTiles?: number,
): Promise<HTMLCanvasElement> {
  const tiles = await loadTiles();
  const maxLen = lines.reduce((m, l) => Math.max(m, l.length), 0);
  const totalW = widthTiles ?? maxLen + 4;
  const totalH = lines.length + 2;

  const canvas = document.createElement('canvas');
  canvas.width = totalW * TILE_SIZE;
  canvas.height = totalH * TILE_SIZE;
  const ctx = canvas.getContext('2d');
  if (!ctx) throw new Error('2d context unavailable');

  const imageData = ctx.createImageData(canvas.width, canvas.height);
  const buf = imageData.data;
  // Fill white.
  for (let i = 0; i < buf.length; i += 4) {
    buf[i] = 255;
    buf[i + 1] = 255;
    buf[i + 2] = 255;
    buf[i + 3] = 255;
  }

  const paste = (tileIdx: number, tx: number, ty: number) => {
    const tile = getTile(tiles, tileIdx);
    const px = tx * TILE_SIZE;
    const py = ty * TILE_SIZE;
    for (let y = 0; y < TILE_SIZE; y++) {
      for (let x = 0; x < TILE_SIZE; x++) {
        const v = tile.pixels[y * TILE_SIZE + x];
        const di = ((py + y) * canvas.width + (px + x)) * 4;
        buf[di] = v;
        buf[di + 1] = v;
        buf[di + 2] = v;
        buf[di + 3] = 255;
      }
    }
  };

  // Top border.
  paste(BORDER_TL, 0, 0);
  for (let x = 1; x < totalW - 1; x++) paste(BORDER_H, x, 0);
  paste(BORDER_TR, totalW - 1, 0);

  // Body rows: vertical borders + space-filled interior + characters.
  for (let row = 0; row < lines.length; row++) {
    const y = row + 1;
    paste(BORDER_V, 0, y);
    for (let x = 1; x < totalW - 1; x++) paste(0x7f, x, y);
    const line = lines[row];
    for (let ci = 0; ci < line.length; ci++) {
      const idx = CHARMAP[line[ci]] ?? 0x7f;
      paste(idx, ci + 1, y);
    }
    paste(BORDER_V, totalW - 1, y);
  }

  // Bottom border.
  paste(BORDER_BL, 0, totalH - 1);
  for (let x = 1; x < totalW - 1; x++) paste(BORDER_H, x, totalH - 1);
  paste(BORDER_BR, totalW - 1, totalH - 1);

  ctx.putImageData(imageData, 0, 0);
  return canvas;
}
